### ISSUE TITLE
Incompatible data type for numpy ops?

### ISSUE THREAD

**Issue by eric-haibin-lin** [apache/incubator-mxnet#16111]
```
>>> mx.np.log(mx.np.array([1], dtype='int'))
array([0], dtype=int64)

>>> np.log(np.array([10]))
array([2.30258509])
```

Many math operators in numpy (such as `log`) do not return integer result. However, `mx.np` does. Is this expected?

**Comment by eric-haibin-lin** [comment]
@reminisce @sxjscience

**Comment by reminisce** [comment]
Yeah, we are already aware of this and has been working on this. It requires us to re-write the kernel of those ops and that would become much less tedious and efforts after we use TVM op module developed by @yzhliu . One catch here is that we may not be able to provide backward compute function for int input types as the backward pass shares the same infer type function as the forward. So for those ops taking integer inputs, we give error message that the gradient calculation is not supported and ask users to use floats as inputs. What do you guys think?

**Comment by sxjscience** [comment]
@reminisce I think it's reasonable.

**Comment by sxjscience** [comment]
Link this with https://github.com/apache/incubator-mxnet/issues/16048
